### PR TITLE
feat: add support for datadog tag propagation

### PR DESCRIPTION
**Pull request by majorgreys**
## Description

Add support for propagating `x-datadog-tags` in HTTP headers. 

Trace tag are added as tags on the trace root span.

Any errors in propagation of trace tags are reported as the `_dd.propagation.error` tag on the trace root span.

## Checklist
- [X] Title must conform to [conventional commit](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional).
- [X] Add additional sections for `feat` and `fix` pull requests.
- [X] Ensure tests are passing for affected code.
- [X] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Motivation

These trace tags will be used to share tracer state across service boundaries. 

## Testing strategy

Automated tests included.

### CODE REVIEW

File: benchmarks/http_propagation_extract/scenario.py [hunk 1]
```diff
@@ -2,8 +2,8 @@
 
 import bm
 
-from ddtrace.propagation import _utils as utils
 from ddtrace.propagation import http
+from ddtrace.propagation import utils
```
Comment (brettlangdon) [inline benchmarks/http_propagation_extract/scenario.py:6]: is this going to fail? 1.x it is `_utils` right?

File: benchmarks/http_propagation_inject/config.yaml [hunk 2]
```diff
@@ -15,3 +16,25 @@ with_priority_and_origin:
   sampling_priority: "1"
   dd_origin: "synthetics"
 
+with_tags:
+  <<: *defaults
+  meta: |
+    {"_dd.p.test": "value", "_dd.p.sample": "value", "will.be": "skipped"}
```
Comment (brettlangdon) [inline benchmarks/http_propagation_inject/config.yaml:22]: this should be a valid tag, since we are not going to blindly inject all `_dd.p.*` tags.

File: ddtrace/propagation/http.py [hunk 3]
```diff
@@ -36,12 +44,16 @@ def _possible_header(header):
     return frozenset([header, get_wsgi_header(header).lower()])
 
 
+HTTP_HEADER_TAGS = "x-datadog-tags"
```
Comment (brettlangdon) [inline ddtrace/propagation/http.py:47]:
```suggestion
_HTTP_HEADER_TAGS = "x-datadog-tags"
```
❓ 🤷🏻

File: ddtrace/propagation/http.py [hunk 4]
```diff
@@ -108,6 +120,35 @@ def _inject(span_context, headers):
         if span_context.dd_origin is not None:
             headers[HTTP_HEADER_ORIGIN] = ensure_text(span_context.dd_origin)
 
+        # Do not try to encode tags if we have already tried and received an error
+        if "_dd.propagation_error" in span_context._meta:
+            return
+
+        # TODO: add whitelist
```
Comment (brettlangdon) [inline ddtrace/propagation/http.py:127]:
```suggestion
        # TODO: add allowlist
```

Comment (brettlangdon) [reply ddtrace/propagation/http.py:127]: 👍🏻 we definitely want this

Comment (brettlangdon) [reply ddtrace/propagation/http.py:127]: we also need to check if this feature is enabled

File: ddtrace/propagation/http.py [hunk 5]
```diff
@@ -423,6 +482,17 @@ def parent_call():
...
+    @staticmethod
+    def _extract_header_value(possible_header_names, headers, default=None):
+        # type: (FrozenSet[str], Dict[str, str], Optional[str]) -> Optional[str]
+        for header in possible_header_names:
+            try:
+                return headers[header]
+            except KeyError:
+                pass
+
+        return default
```
Comment (brettlangdon) [inline ddtrace/propagation/http.py:494]: this can be removed.

File: ddtrace/propagation/http.py [hunk 6]
```diff
@@ -131,6 +172,19 @@ def _extract(headers):
             POSSIBLE_HTTP_HEADER_ORIGIN,
             headers,
         )
+        meta = None
+        tags_value = HTTPPropagator._extract_header_value(
```
Comment (brettlangdon) [inline ddtrace/propagation/http.py:176]:
```suggestion
        tags_value = _extract_header_value(
```

File: releasenotes/notes/disable-internal-tag-propagation-dff3e799fb056584.yaml [hunk 7]
```diff
@@ -1,4 +0,0 @@
----
-other:
-  - |
-    tracing/internal: disable Datadog internal tag propagation
```
Comment (brettlangdon) [inline releasenotes/notes/disable-internal-tag-propagation-dff3e799fb056584.yaml:4]: 🤔 why is this here?

**Review by brettlangdon — CHANGES_REQUESTED** [review]

File: ddtrace/propagation/http.py [hunk 8]
```diff
@@ -108,6 +120,35 @@ def _inject(span_context, headers):
...
+            # DEV: encoding will fail if the key or value are not `str`
+            key = ensure_str(key)
+            if key.startswith("_dd.p."):
+                tags_to_encode[key] = ensure_str(value)
+
+        if tags_to_encode:
+            encoded_tags = None
+
+            try:
+                encoded_tags = encode_tagset_values(tags_to_encode)
```
Comment (P403n1x87) [inline ddtrace/propagation/http.py:140]:
```suggestion
                headers[HTTP_HEADER_TAGS] = encode_tagset_values(tags_to_encode)
```

File: ddtrace/propagation/http.py (hunk 4 above)
Comment (majorgreys) [reply ddtrace/propagation/http.py:127]: @brettlangdon what "feature" do you mean here? It would seem that in this PR we should only support an allowed list of tags, specifically just `_dd.p.dm` and `_dd.propagation.error`.

Comment (brettlangdon) [reply ddtrace/propagation/http.py:127]: propagating tags via headers.

```
DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH: default=128

If the value is 0, we can consider the propagating feature is disabled. We still should parse the incoming x-datadog-tags header even if the propagation was disabled.

Trace tags propagation is disabled: set _dd.propagation_error:disabled  in the root span
```

File: ddtrace/propagation/http.py [hunk 9]
```diff
@@ -120,24 +120,24 @@ def _inject(span_context, headers):
         if span_context.dd_origin is not None:
             headers[HTTP_HEADER_ORIGIN] = ensure_text(span_context.dd_origin)
 
+        # Skip injecting tracer tags
+        if not config._propagation_datadog_tags_enabled:
```
Comment (brettlangdon) [inline ddtrace/propagation/http.py:124]: We need to add a meta tag to the trace saying we didn't propagate because it is disabled.

File: ddtrace/propagation/http.py [hunk 10]
```diff
@@ -103,6 +101,8 @@ class _DatadogMultiHeader:
       - ``x-datadog-origin`` optional name of origin Datadog product which initiated the request
     """
 
+    _PROPAGATED_TAGS = frozenset(["_dd.p.dm", "_dd.propagation_error"])
```
Comment (brettlangdon) [inline ddtrace/propagation/http.py:104]:
```suggestion
    _PROPAGATED_TAGS = frozenset(["_dd.p.dm"])
```

`_dd.propagation_error` is only added to the span meta, it doesn't get propagated (... right?)

File: ddtrace/propagation/http.py [hunk 11]
```diff
@@ -120,24 +120,24 @@ def _inject(span_context, headers):
...
-        for key, value in span_context._meta.items():
-            # DEV: encoding will fail if the key or value are not `str`
-            key = ensure_str(key)
-            if key.startswith("_dd.p."):
-                tags_to_encode[key] = ensure_str(value)
+        # Only propagate tags that are allowed for propagation
+        tags_to_encode = {  # type: Dict[str, str]
+            k: v
+            for k, v in ((ensure_str(k), v) for k, v in span_context._meta.items())
+            if k in _DatadogMultiHeader._PROPAGATED_TAGS
```
Comment (brettlangdon) [inline ddtrace/propagation/http.py:135]: I wonder if we should reverse this:

```python
tags_to_encode = {
    k: span_context._meta[k]
    for k in _DatadogMultiHeader._PROPAGATED_TAGS
    if k in span_context._meta
}
```

Should consistently be less iterations?

Comment (brettlangdon) [reply ddtrace/propagation/http.py:135]: (micro optimization?)

File: ddtrace/propagation/http.py [hunk 12]
```diff
@@ -182,10 +181,17 @@ def _extract(headers):
             # Do not fail if the tags are malformed
             try:
                 # We get a Dict[str, str], but need it to be Dict[Union[str, bytes], str] (e.g. _MetaDictType)
-                meta = cast(Dict[Union[str, bytes], str], decode_tagset_string(tags_value))
+                meta = cast(
+                    Dict[Union[str, bytes], str],
+                    {k: v for (k, v) in decode_tagset_string(tags_value) if k in _DatadogMultiHeader._PROPAGATED_TAGS},
```
Comment (brettlangdon) [inline ddtrace/propagation/http.py:186]: we can take them all, we don't need to filter what we extract.

if it is already in the http headers, there shouldn't be much issue with adding to the span meta.

File: ddtrace/propagation/http.py [hunk 13]
```diff
@@ -182,10 +181,17 @@ def _extract(headers):
...
+                )
             except TagsetDecodeError:
                 log.debug("failed to decode x-datadog-tags: %r", tags_value, exc_info=True)
 
+        # Parse tags even if propagation is disabled
+        if not config._propagation_datadog_tags_enabled:
+            meta = None
```
Comment (brettlangdon) [inline ddtrace/propagation/http.py:193]: I think this can go away? it doesn't match the comment.

File: docs/configuration.rst [hunk 14]
```diff
@@ -134,7 +134,6 @@ below:
 
        .. _dd-trace-writer-max-payload-size-bytes:
    * - ``DD_TRACE_WRITER_MAX_PAYLOAD_SIZE_BYTES``
-     - Int
```
Comment (brettlangdon) [inline docs/configuration.rst:137]: ?

File: ddtrace/propagation/http.py (hunk 11 above)
Comment (majorgreys) [reply ddtrace/propagation/http.py:135]: we need to do the `ensure_str` before doing the lookup

File: ddtrace/propagation/http.py (hunk 13 above)
Comment (majorgreys) [reply ddtrace/propagation/http.py:193]: We don't skip the previous decoding of tagset and reset the meta if this feature is disabled. I can clarify that in the comment.

File: ddtrace/propagation/http.py (hunk 11 above)
Comment (P403n1x87) [reply ddtrace/propagation/http.py:134]: I would `ensure_str` for the value only if the test on `k` is `True`. WDYT?

File: ddtrace/propagation/http.py (hunk 9 above)
Comment (Kyle-Verhoog) [reply ddtrace/propagation/http.py:123]: should we only do this if there are tags and propagation is disabled? seems unnecessary to add it when there are no tags received?

File: ddtrace/propagation/http.py [hunk 15]
```diff
@@ -132,6 +172,32 @@ def _extract(headers):
...
+                        k: v
+                        for (k, v) in decode_tagset_string(tags_value).items()
+                        if k in _DatadogMultiHeader._PROPAGATED_TAGS
+                    },
+                )
+            except TagsetDecodeError:
+                log.debug("failed to decode x-datadog-tags: %r", tags_value, exc_info=True)
+            finally:
+                # Datadog propagation should still attempt to decode even if tag
+                # propagation is disabled
```
Comment (Kyle-Verhoog) [inline ddtrace/propagation/http.py:197]: why? could this be a security vulnerability to ddos a webserver?

File: ddtrace/propagation/http.py [hunk 16]
```diff
@@ -423,6 +494,17 @@ def parent_call():
         if PROPAGATION_STYLE_B3_SINGLE_HEADER in config._propagation_style_inject:
             _B3SingleHeader._inject(span_context, headers)
 
+    @staticmethod
+    def _extract_header_value(possible_header_names, headers, default=None):
+        # type: (FrozenSet[str], Dict[str, str], Optional[str]) -> Optional[str]
+        for header in possible_header_names:
+            try:
```
Comment (Kyle-Verhoog) [inline ddtrace/propagation/http.py:501]:
```suggestion
            if header in headers:
```

```python
In [8]: keys = ["asdf", "ASDF", "fdsa", "FDSA"]

In [9]: def fn_exc(d, default):
   ...:     for k in keys:
   ...:         try:
   ...:             return d[k]
   ...:         except KeyError:
   ...:             pass
   ...:     return default
   ...: 


In [10]: def fn_get(d, default):
    ...:     for k in keys:
    ...:         if k in d:
    ...:             return d[k]
    ...:     return default
    ...: 

In [11]: %timeit fn_exc({}, 123)
489 ns ± 2.35 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [12]: %timeit fn_get({}, 123)
130 ns ± 0.45 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [13]: %timeit fn_exc({"asdf": 321}, 123)
108 ns ± 0.524 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [14]: %timeit fn_get({"asdf": 321}, 123)
114 ns ± 0.175 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

some pretty big savings if we avoid the exception handling and there are many misses

File: ddtrace/propagation/http.py [hunk 17]
```diff
@@ -423,6 +494,17 @@ def parent_call():
...
+                return headers[header]
+            except KeyError:
```
Comment (Kyle-Verhoog) [inline ddtrace/propagation/http.py:503]:
```suggestion
```

File: ddtrace/propagation/http.py [hunk 18]
```diff
@@ -423,6 +494,17 @@ def parent_call():
...
+                pass
```
Comment (Kyle-Verhoog) [inline ddtrace/propagation/http.py:504]:
```suggestion
```

File: tests/tracer/test_propagation.py [hunk 19]
```diff
@@ -20,19 +20,223 @@
 from ddtrace.propagation.http import _HTTP_HEADER_B3_SINGLE
```
Comment (Kyle-Verhoog) [inline tests/tracer/test_propagation.py:20]: we should probably have a test case that creates a trace with multiple spans. It looks like by this implementation that the propagated tags will be applied to each span but this isn't enforced or confirmed by any tests. Maybe a snapshot test would be useful for ensuring this?

File: ddtrace/settings/config.py [hunk 20]
```diff
@@ -193,6 +193,12 @@ def __init__(self):
             "DD_TRACE_PROPAGATION_STYLE_INJECT", default=PROPAGATION_STYLE_DATADOG
         )
 
+        # Datadog tracer tags propagation
+        self._propagation_datadog_tags_max_length = max(
```
Comment (Kyle-Verhoog) [inline ddtrace/settings/config.py:197]: Should probably fail if the value is < 0 rather than coerce to 0

**Review by Kyle-Verhoog — COMMENTED** [review]
> Only _dd.p.dm is allowed as a propagated trace tag.

What does this mean? We only consider a single tag key when injecting, extracting?

Can we get a little more context about the implementation in the PR description? Are these tags applied to all spans or just the local top level? Is the feature enabled by default? Any performance implications?

File: tests/tracer/test_propagation.py [hunk 21]
```diff
@@ -20,19 +20,216 @@
...
+        assert span.trace_id == 1234
+        assert span.parent_id == 5678
+        assert span.context.sampling_priority == 1
+        assert span.context.dd_origin == "synthetics"
+        assert span.context._meta == {
+            "_dd.origin": "synthetics",
+            "_dd.p.dm": "value",
+            "any": "tag",
+        }
+        # TODO: Should x-datadog-tags be in shallow copy?
```
Comment (majorgreys) [inline tests/tracer/test_propagation.py:166]: @Kyle-Verhoog In adding a multispan trace as you suggested, we see that the activated context (with its `_meta`) is shallow copied when a child span is started. I had assumed these "trace level" metadata would not be copied to all spans but I realize now that it would have to given how we inject in the propagators using the context on the current span rather than the local root span. Should we retain this behavior for `x-datadog-tags`?

File: ddtrace/propagation/http.py (hunk 16 above)
Comment (majorgreys) [reply ddtrace/propagation/http.py:512]: I will open a separate PR for this optimization.

File: tests/tracer/test_propagation.py (hunk 21 above)
Comment (majorgreys) [reply tests/tracer/test_propagation.py:166]: Still need to add snapshot test for this. to integration test.

**Review by Kyle-Verhoog — COMMENTED** [review]
looking pretty good to me, just a few small things + the other open comments

File: ddtrace/propagation/http.py (hunk 15 above)
Comment (majorgreys) [reply ddtrace/propagation/http.py:211]: This is by design as the expected behavior is to propagate tags forward even if those tags are not supported yet by a library.

File: tests/tracer/test_propagation.py [hunk 22]
```diff
@@ -20,19 +20,242 @@
...
 
 from ..utils import override_global_config
 
 
 NOT_SET = object()
 
 
+def test_inject(tracer):
+    # We will only pass along `_dd.p.dm` tags
+    meta = {"_dd.p.dm": "value", "_dd.p.other": "value"}
```
Comment (majorgreys) [inline tests/tracer/test_propagation.py:33]: Need to verify that this is the expected behavior given that we still parse the tag that is not in the allowed list.

File: tests/integration/test_propagation.py [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+        tracer.configure(**kwargs)
+        yield tracer
+        tracer.shutdown()
+
+
+@pytest.mark.parametrize(
+    "trace_tags_and_span_check",
+    [
+        ("_dd.p.dm=value,any=tag", lambda s: s["meta"].get("_dd.p.dm") == "value" and s["meta"].get("any") == "tag"),
+        ("_dd.p.upstream_services=foo", lambda s: "meta" not in s or "_dd.p.upstream_services" not in s["meta"]),
```
Comment (majorgreys) [inline tests/integration/test_propagation.py:43]: This complexity could be avoided if we just broke this test out into two.

File: tests/integration/test_propagation.py [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+    ]
+)
+def tracer(request):
+    global_config = request.param["global_config"]
+    partial_flush_enabled = request.param.get("partial_flush_enabled", False)
+    partial_flush_min_spans = request.param.get("partial_flush_min_spans", None)
+    with override_global_config(request.param):
+        tracer = Tracer()
+        out = StringIO()
+        kwargs = dict(writer=LogWriter(out=out))
```
Comment (majorgreys) [inline tests/integration/test_propagation.py:29]: This can be replaced with the test agent for snapshots.

File: ddtrace/propagation/http.py [hunk 25]
```diff
@@ -130,6 +178,40 @@ def _extract(headers):
...
+            except TagsetDecodeError:
+                meta = {
+                    "_dd.propagation_error": "decoding_error",
+                }
+                log.debug("failed to decode x-datadog-tags: %r", tags_value, exc_info=True)
+            finally:
+                # Datadog propagation should still attempt to decode even if tag
+                # propagation is disabled
+                if not config._x_datadog_tags_enabled:
+                    meta = None
```
Comment (majorgreys) [inline ddtrace/propagation/http.py:213]: We don't want to replace `meta` if an exception has been raised, no need for cleanup.

```suggestion
```

File: ddtrace/propagation/http.py [hunk 26]
```diff
@@ -87,8 +98,17 @@ class _DatadogMultiHeader:
...
       - ``x-datadog-origin`` optional name of origin Datadog product which initiated the request
+      - ``x-datadog-tags`` optional tracer tags
+
+    Restrictions:
+
+      - All tag-value pairs in `x-datadog-tags` are extracted from incoming requests.
+      - An allow list restricts ``x-datadog-tags`` injected into outgoing requests.
     """
 
+    _X_DATADOG_TAGS_REJECT = frozenset(["_dd.p.upstream_services"])
```
[the remark this answers is not in the document]
Comment (majorgreys) [reply ddtrace/propagation/http.py:109]: Part of spec

File: ddtrace/propagation/http.py (hunk 15 above)
Comment (Kyle-Verhoog) [reply ddtrace/propagation/http.py]: Do we have to decode if that is the case? We should be able to just forward the encoded payload along without decoding, right?

File: ddtrace/propagation/http.py [hunk 27]
```diff
@@ -106,6 +126,34 @@ def _inject(span_context, headers):
...
+        # Do not try to encode tags if we have already tried and received an error
+        if "_dd.propagation_error" in span_context._meta:
+            return
+
+        # Only propagate tags that are allowed for propagation
+        tags_to_encode = {
+            k: ensure_str(v)
+            for k, v in ((ensure_str(k), v) for k, v in span_context._meta.items())
+            if k in _DatadogMultiHeader._X_DATADOG_TAGS_ALLOW
+        }  # type: Dict[str, str]
```
[the remark this answers is not in the document]
Comment (majorgreys) [reply ddtrace/propagation/http.py]: I put a comment in the code for why we need to coerce.

If we want to avoid this, we would need to change the typing of the meta dictionary in `Context` vs `Span`.

File: tests/tracer/test_propagation.py (hunk 22 above)
Comment (majorgreys) [reply tests/tracer/test_propagation.py:32]: We do not have an allow list for extract/inject.

File: ddtrace/propagation/http.py (hunk 15 above)
Comment (majorgreys) [reply ddtrace/propagation/http.py]: We do need to decode as we need to report if there are any errors in decoding.

**Review by brettlangdon — COMMENTED** [review]
few nits, otherwise lgtm

File: ddtrace/propagation/http.py [hunk 28]
```diff
@@ -106,6 +133,38 @@ def _inject(span_context, headers):
...
+        }  # type: Dict[Text, Text]
+
+        if tags_to_encode:
+            try:
+                headers[_HTTP_HEADER_TAGS] = encode_tagset_values(
+                    tags_to_encode, max_size=config._x_datadog_tags_max_length
+                )
+            except TagsetMaxSizeEncodeError:
+                # We hit the max size allowed, add a tag to the context to indicate this happened
+                span_context._meta["_dd.propagation_error"] = "inject_max_size"
```
[the remark this answers is not in the document]
Comment (majorgreys) [reply ddtrace/propagation/http.py:159]: The decision made in the spec for this feature is not to extract or inject the partial set of tags.

File: ddtrace/propagation/http.py [hunk 29]
```diff
@@ -106,6 +133,36 @@ def _inject(span_context, headers):
...
+        # Do not try to encode tags if we have already tried and received an error
+        if "_dd.propagation_error" in span_context._meta:
+            return
+
+        # Only propagate trace tags which means ignoring the _dd.origin
+        tags_to_encode = {
+            # DEV: Context._meta is a _MetaDictType but we need Dict[str, str]
+            ensure_str(k): ensure_str(v)
+            for k, v in span_context._meta.items()
+            if _DatadogMultiHeader.is_valid_datadog_trace_tag_key(k)
```
Comment (brettlangdon) [inline ddtrace/propagation/http.py:149]:
```suggestion
            if _DatadogMultiHeader._is_valid_datadog_trace_tag_key(k)
```

File: ddtrace/propagation/http.py [hunk 30]
```diff
@@ -130,6 +187,32 @@ def _extract(headers):
...
+            default="",
+        )
+        if tags_value:
+            # Do not fail if the tags are malformed
+            try:
+                meta = {
+                    k: v
+                    for (k, v) in decode_tagset_string(tags_value).items()
+                    if k not in _DatadogMultiHeader._X_DATADOG_TAGS_EXTRACT_REJECT
+                    if _DatadogMultiHeader.is_valid_datadog_trace_tag_key(k)
```
Comment (brettlangdon) [inline ddtrace/propagation/http.py:203]:
```suggestion
                    if _DatadogMultiHeader._is_valid_datadog_trace_tag_key(k)
```

File: tests/integration/test_propagation.py [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+
+
+pytestmark = pytest.mark.skipif(AGENT_VERSION != "testagent", reason="Tests only compatible with a testagent")
+
+
+@pytest.fixture(
+    params=[
+        dict(global_config=dict()),
+        dict(
+            global_config=dict(_x_datadog_tags_max_length="0", _x_datadog_tags_enabled=False),
```
[the remark this answers is not in the document]
Comment (brettlangdon) [reply tests/integration/test_propagation.py:17]: Wait, I am confused, you have it as a str 😕

File: tests/tracer/test_propagation.py [hunk 32]
```diff
@@ -20,19 +20,240 @@
...
+    tracer.context_provider.activate(ctx)
+    with tracer.trace("global_root_span") as span:
+        headers = {}
+        HTTPPropagator.inject(span.context, headers)
+
+        assert int(headers[HTTP_HEADER_TRACE_ID]) == span.trace_id
+        assert int(headers[HTTP_HEADER_PARENT_ID]) == span.span_id
+        assert int(headers[HTTP_HEADER_SAMPLING_PRIORITY]) == span.context.sampling_priority
+        assert headers[HTTP_HEADER_ORIGIN] == span.context.dd_origin
+        assert ("_dd.p.dm=value,_dd.p.other=value" in headers[_HTTP_HEADER_TAGS]) or (
```
Comment (brettlangdon) [inline tests/tracer/test_propagation.py:43]: Below you do this, any reason to be different here?

```python
        # The ordering is non-deterministic, so compare as a list of tags
        tags = set(headers[_HTTP_HEADER_TAGS].split(","))
        assert tags == set(["_dd.p.dm=bytes"])
```

I actually don't care that much, curiosity, feel free to resolve this without any action.

**Review by brettlangdon — APPROVED** [review]

File: ddtrace/settings/config.py [hunk 33]
```diff
@@ -193,6 +193,18 @@ def __init__(self):
             "DD_TRACE_PROPAGATION_STYLE_INJECT", default=PROPAGATION_STYLE_DATADOG
         )
 
+        # Datadog tracer tags propagation
+        x_datadog_tags_max_length = int(os.getenv("DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH", default=128))
```
Comment (wantsui) [inline ddtrace/settings/config.py:197]: Based on the spec, should the default here be **512** instead of **128**?

The file in `ddtrace/internal/_tagset.pyx` also mentions that the default is **512**.

File: docs/configuration.rst [hunk 34]
```diff
@@ -204,6 +204,12 @@ below:
        Example: ``DD_TRACE_PROPAGATION_STYLE_INJECT="datadog,b3"`` to inject both ``x-datadog-*`` and ``x-b3-*``
        headers into outbound requests.
 
+       .. _dd-trace-x-datadog-tags-max-length:
+   * - ``DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH``
+     - Integer
+     - 128
```
Comment (wantsui) [inline docs/configuration.rst:210]: Should the default here be **512** instead of **128** (See comment in `config.py`)?

File: ddtrace/propagation/http.py [hunk 35]
```diff
@@ -130,6 +187,32 @@ def _extract(headers):
...
+                    k: v
+                    for (k, v) in decode_tagset_string(tags_value).items()
+                    if k not in _DatadogMultiHeader._X_DATADOG_TAGS_EXTRACT_REJECT
+                    if _DatadogMultiHeader._is_valid_datadog_trace_tag_key(k)
+                }
+            except TagsetMaxSizeDecodeError:
+                meta = {
+                    "_dd.propagation_error": "extract_max_size",
+                }
+                log.warning("failed to decode x-datadog-tags", exc_info=True)
```
Comment (wantsui) [inline ddtrace/propagation/http.py:209]: Would it be useful to log the size of the tag if this error ever gets thrown?

File: docs/configuration.rst [hunk 36]
```diff
@@ -204,6 +204,12 @@ below:
...
+     - If the value is 0, Datadog trace tags propagation is disabled.
```
Comment (wantsui) [inline docs/configuration.rst:211]: It might be useful to specify in this configuration table that only _positive_ integers can be used, since an error gets thrown if someone uses a negative number in `config.py`. 

Side question: is there a case where someone may want to increase the default?

File: tests/tracer/test_settings.py [hunk 37]
```diff
@@ -320,3 +322,24 @@ def test_environment_header_tags():
...
+    yield
+
+
+@pytest.mark.parametrize(
+    "env,expected,raises",
+    (
+        (dict(), (128, True), no_exceptions_raised()),
+        (dict(DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH="0"), (0, False), no_exceptions_raised()),
+        (dict(DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH="256"), (256, True), no_exceptions_raised()),
+        (dict(DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH="-1"), None, pytest.raises(ValueError)),
```
Comment (wantsui) [inline tests/tracer/test_settings.py:338]: Should there be a test for an extremely large int and other defaults to see whether it throws an error or not? 

Example values: `1000` and `512`?

**Review by wantsui — CHANGES_REQUESTED** [review]
This feature looks really interesting! My comments are mainly on the expected values for `DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH` and how it might be used.

File: ddtrace/settings/config.py (hunk 33 above)
Comment (majorgreys) [reply ddtrace/settings/config.py:197]: Good catch!

File: ddtrace/propagation/http.py (hunk 35 above)
Comment (majorgreys) [reply ddtrace/propagation/http.py:209]: added this to the exception message

File: releasenotes/notes/add-datadog-trace-tag-propagation-f69736976107f1c2.yaml [hunk 38]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    Add support for Datadog trace tag propagation
```
Comment (wantsui) [inline releasenotes/notes/add-datadog-trace-tag-propagation-f69736976107f1c2.yaml:4]: Should the release note explain what new environment variables are introduced with this PR?

**Review by wantsui — CHANGES_REQUESTED** [review]
I have one last comment about `DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH`. In general, is this supposed to be a value between 0 to 512?

File: docs/configuration.rst [hunk 39]
```diff
@@ -204,6 +204,12 @@ below:
        Example: ``DD_TRACE_PROPAGATION_STYLE_INJECT="datadog,b3"`` to inject both ``x-datadog-*`` and ``x-b3-*``
        headers into outbound requests.
 
+       .. _dd-trace-x-datadog-tags-max-length:
+   * - ``DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH``
+     - Integer
+     - 512
+     - Maximum length (non-negative) of ``x-datadog-tags`` by Datadog propagation style. If 0, propagation of ``x-datadog-tags`` is disabled.
```
[the remark this answers is not in the document]
Comment (majorgreys) [reply docs/configuration.rst]: The max value is in the table already

File: releasenotes/notes/add-datadog-trace-tag-propagation-f69736976107f1c2.yaml (hunk 38 above)
Comment (majorgreys) [reply releasenotes/notes/add-datadog-trace-tag-propagation-f69736976107f1c2.yaml:4]: I don't think this is needed in the release notes. That should be included in documentation for this feature when that lands.

File: docs/configuration.rst [hunk 40]
```diff
@@ -204,6 +204,12 @@ below:
        Example: ``DD_TRACE_PROPAGATION_STYLE_INJECT="datadog,b3"`` to inject both ``x-datadog-*`` and ``x-b3-*``
        headers into outbound requests.
 
+       .. _dd-trace-x-datadog-tags-max-length:
+   * - ``DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH``
+     - Integer
+     - 512
+     - The maximum length (non-negative) of ``x-datadog-tags`` header allowed in the Datadog propagation style. If 0, propagation of ``x-datadog-tags`` is disabled.
```
Comment (wantsui) [inline docs/configuration.rst:211]: [optional suggestion to make it clear what values can be set]
```git suggest

     - The maximum length (non-negative) of ``x-datadog-tags`` header allowed in the Datadog propagation style. Must be a value between 0 to 512. If 0, propagation of ``x-datadog-tags`` is disabled.
```

**Review by wantsui — DISMISSED** [review]
LGTM 👍

File: docs/configuration.rst (hunk 40 above)
Comment (majorgreys) [inline docs/configuration.rst:211]:
```suggestion
     - The maximum length of ``x-datadog-tags`` header allowed in the Datadog propagation style. Must be a value between 0 to 512. If 0, propagation of ``x-datadog-tags`` is disabled.
```

**Review by brettlangdon — DISMISSED** [review]

**Review by P403n1x87 — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]
